### PR TITLE
Show up task summary when user clic task row

### PR DESCRIPTION
**Pull request by Angatupyry**
## What's new
Popup dialog to show task summary

- Removed microapps (Task logs and Task details) from the top right of the task section

- Added a task summary with the most relevant data
- Task details and Task logs are now displayed by clicking a button in the summary component.

[chrome-capture-2023-3-17.webm](https://user-images.githubusercontent.com/37310205/232598385-1737d969-1b9c-4d8a-932e-d352cc1446a3.webm)

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import { AppEvents } from '../app-events';
+import { RmfAppContext } from '../rmf-app';
+import { TaskInfo } from 'react-components';
+import { UserProfileContext } from 'rmf-auth';
+import { Enforcer } from '../permissions';
+import { TaskLogs } from './task-logs';
+
+export interface TableDataGridState {
+  task: TaskState | null;
+  open: boolean;
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:25]: Similar to the comment in `task-summary`, we can probably do away with this prop, by setting a state that is default true,
```
const [isOpen, setIsOpen] = React.useState(true);
```
and gets set to false in the `onClose` dialog prop

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+export function TaskInspector({ task, open, onClose }: TableDataGridState): JSX.Element {
+  const theme = useTheme();
+  const rmf = React.useContext(RmfAppContext);
+  const appController = React.useContext(AppControllerContext);
+
+  const [taskState, setTaskState] = React.useState<TaskState | null>(null);
+  const [taskLogs, setTaskLogs] = React.useState<TaskEventLog | null>(null);
+  const [waypoint, setWaypoint] = React.useState({
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:36]: I think we should separate this into `startWaypoint` and `finishWaypoint`, since we are not encapsulating that much information in this object

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+          selectedTask ? rmf.getTaskStateObs(selectedTask.booking.id) : of(null),
+        ),
+      )
+      .subscribe(setTaskState);
+    return () => sub.unsubscribe();
+  }, [rmf]);
+
+  React.useEffect(() => {
+    AppEvents.taskSelect.next(task);
+  }, [task]);
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:118]: Instead of setting up this workaround chain of `AppEvents.taskSelect`, I think we can make use of https://github.com/open-rmf/rmf-web/blob/main/packages/dashboard/src/components/rmf-app/rmf-ingress.ts#L201, subscribing directly to the ID of the `task` prop, and retrieving the logs using the same API call you used

File: packages/react-components/lib/tasks/task-info.tsx [hunk 4]
```diff
@@ -26,108 +26,25 @@ function InfoValue({ children }: React.PropsWithChildren<unknown>) {
...
   return (
     <StyledDiv>
       <Typography variant="h6" style={{ textAlign: 'center' }} gutterBottom>
-        {task.booking.id}
+        {title ? title : task.booking.id}
       </Typography>
       <Divider />
       <div style={{ marginBottom: theme.spacing(1) }}></div>
       <InfoLine>
         <span>State:</span>
```
Comment (aaronchongth) [inline packages/react-components/lib/tasks/task-info.tsx:45]:
```suggestion
        <span>Status:</span>
```

File: packages/react-components/lib/tasks/task-info.tsx [hunk 5]
```diff
@@ -26,108 +26,25 @@ function InfoValue({ children }: React.PropsWithChildren<unknown>) {
...
         <InfoValue>{task.status || 'unknown'}</InfoValue>
       </InfoLine>
```
Comment (aaronchongth) [inline packages/react-components/lib/tasks/task-info.tsx:47]: Referring back to figma, I believe we may have shuffled up some of the intended fields, `Start waypoint` and `Finish waypoint` should be below this `Status`/`State` field within the `TaskInfo` components. Migrating the waypoint extraction logic down to this component will probably make `TaskInspector` much cleaner too

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+              </DialogTitle>
+            </Grid>
+            <Grid container>
+              <Grid item xs={2} ml={7}>
+                <Typography>Status:</Typography>
+              </Grid>
+              <Grid item xs={3}>
+                <Typography align="right">{task?.status}</Typography>
+              </Grid>
+            </Grid>
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:172]: This is probably redundant, since we have the status listed under `TaskInfo`

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            <Grid container>
+              <Grid item xs={2} ml={7}>
+                <Typography>Finish Waypoint:</Typography>
+              </Grid>
+              <Grid item xs={3}>
+                <Typography align="right">{waypoint.finish}</Typography>
+              </Grid>
+            </Grid>
+          </Grid>
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:189]: As mentioned in a comment below, we should migrate these fields down to `TaskInfo`, based on our initial designs on figma

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+          onClose={onClose}
+          aria-labelledby="alert-dialog-title"
+          aria-describedby="alert-dialog-description"
+          fullWidth
+          maxWidth="md"
+        >
+          <Grid container spacing={2}>
+            <Grid item xs={12}>
+              <DialogTitle id="scroll-dialog-title" align="center">
+                {task?.booking.id}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:162]:
```suggestion
                Task: {task?.booking.id}
```

**Review by aaronchongth — CHANGES_REQUESTED** [review]
This looks amazing! Thanks for putting this together, @Angatupyry, it's really coming together now.

I've some review comments and some suggestions that may help us further clean up the implementation. Let me know what you think!

File: packages/dashboard/src/components/tasks/task-summay.tsx
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-summay.tsx:1]: > `task-summary.tsx`

Thanks for reviewing this typo!

File: packages/dashboard/src/components/tasks/task-summay.tsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    textField: {
+      background: theme.palette.background.default,
+      '&:hover': {
+        backgroundColor: theme.palette.background.default,
+      },
+    },
+  }),
+);
+
+const getTaskProgress = (task: TaskState | null): number => {
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: > We should get the task progress using the estimated remaining time, like this implementation https://github.com/open-rmf/rmf-web/blob/main/packages/dashboard/src/components/robots/robot-info-app.tsx#L45-L58

@aaronchongth do you agree with show progress value only if we can calculate it or if it is undefined should we show a message like "Can't calculate task progress"?

File: packages/dashboard/src/components/tasks/task-summay.tsx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+  const getTaskMessage = (task: TaskState | null) => {
+    if (!task || !task.phases || !task.active) {
+      return "Can't get task message";
+    }
+
+    if (Object.values(task.phases)[task.active - 1]) {
+      const message = Object.values(task.phases)[task.active - 1].detail;
+
+      return message ? message : "Can't get task message";
+    }
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: > Looks like we might be missing an else case here

You're right!
Thanks!

File: packages/dashboard/src/components/tasks/task-summay.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: I think we can skip the progress bar component if we can't calculate it. A console log message mentioning this is the case would be nice

**Comment by aaronchongth** [comment]
Sorry I forgot to mention this in my main review comment.

I noticed that the `TaskSummary` dialog is not responsive as the task changes, while `TaskInspector` is. It would be nice if we could make the `TaskSummary` subscribe to the task state using `getTaskStateObs` as well, and have the information (colors, phase information, etc) updated over time.

File: packages/dashboard/src/components/tasks/task-inspector.tsx (hunk 1 above)
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-inspector.tsx:25]: Same answer as above in task summary

File: packages/react-components/lib/tasks/task-info.tsx (hunk 5 above)
Comment (Angatupyry) [reply packages/react-components/lib/tasks/task-info.tsx:80]: Perfect!
Nice suggestion!

File: packages/dashboard/src/components/tasks/task-inspector.tsx (hunk 3 above)
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-inspector.tsx]: @aaronchongth the solution I did now make sense? Please correct me if my solution is not in line with your suggestion.

**Comment by Angatupyry** [comment]
> Sorry I forgot to mention this in my main review comment.
> 
> I noticed that the `TaskSummary` dialog is not responsive as the task changes, while `TaskInspector` is. It would be nice if we could make the `TaskSummary` subscribe to the task state using `getTaskStateObs` as well, and have the information (colors, phase information, etc) updated over time.

Please take a look at the solution I did

File: packages/dashboard/src/components/tasks/task-summay.tsx [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+  const getTaskMessage = (task: TaskState | null) => {
+    if (!task || !task.phases || !task.active) {
+      return "Can't get task message";
+    }
+
+    if (Object.values(task.phases)[task.active - 1]) {
+      const message = Object.values(task.phases)[task.active - 1].detail;
+
+      return message ? message : "Can't get task message";
```
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: I returned it without casting it to string because I would first like to know your opinion about the image I showed you above. Regardless of that, please see if this solution makes sense.

File: packages/dashboard/src/components/tasks/task-summay.tsx [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+    default:
+      return base.palette.background.default;
+  }
+};
+
+export interface TaskSummaryProps {
+  stopShowing: () => void;
+  task: TaskState | null;
+  show: boolean;
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: Hmm, but we will not require multiple `TaskSummary` dialogs to exist at the same time, so this flag doesn't do much, since we are already passing in a callback to be called during `onClose`, we can lump in any closing or toggling logic in that callback.

I'll suggest in a new set of review comments how this can work. Let me know what you think.

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    default:
+      return base.palette.background.default;
+  }
+};
+
+export interface TaskSummaryProps {
+  onClose: () => void;
+  task: TaskState | null;
+  show: boolean;
+}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:72]:
```suggestion
export interface TaskSummaryProps {
  onClose: () => void;
  task: TaskState | null;
}
```

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+
+export const TaskSummary = React.memo((props: TaskSummaryProps) => {
+  const classes = useStyles();
+  const rmf = React.useContext(RmfAppContext);
+
+  const { onClose, show, task } = props;
+
+  const [openTaskDetailsLogs, setOpenTaskDetailsLogs] = React.useState(false);
+  const [taskState, setTaskState] = React.useState<TaskState | null>(null);
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:81]:
```suggestion
  const [taskState, setTaskState] = React.useState<TaskState | null>(null);
  const [isOpen, setIsOpen] = React.useState(true);
```

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+  return (
+    <Dialog
+      PaperProps={{
+        style: {
+          backgroundColor: setTaskDialogColor(taskState?.status),
+          boxShadow: 'none',
+        },
+      }}
+      open={show}
+      onClose={onClose}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:177]:
```suggestion
      open={isOpen}
      onClose={() => {
        setIsOpen(false);
        onClose();
      }}
```

File: packages/dashboard/src/components/tasks/tasks-app.tsx [hunk 16]
```diff
@@ -247,6 +253,13 @@ export const TasksApp = React.memo(
             </Grid>
           </Grid>
           <input type="file" style={{ display: 'none' }} ref={uploadFileInputRef} />
+          {openTaskSummary && (
+            <TaskSummary
+              onClose={() => setOpenTaskSummary(!openTaskSummary)}
+              task={selectedTask}
+              show={openTaskSummary}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/tasks-app.tsx:260]:
```suggestion
              onClose={() => setOpenTaskSummary(false)}
              task={selectedTask}
```

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+import { TaskInfo } from 'react-components';
+import { UserProfileContext } from 'rmf-auth';
+import { Enforcer } from '../permissions';
+import { TaskLogs } from './task-logs';
+
+export interface TableDataGridState {
+  task: TaskState | null;
+  open: boolean;
+  onClose: () => void;
+}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:26]:
```suggestion
export interface TableDataGridState {
  task: TaskState | null;
  onClose: () => void;
}
```

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+export function TaskInspector({ task, open, onClose }: TableDataGridState): JSX.Element {
+  const theme = useTheme();
+  const rmf = React.useContext(RmfAppContext);
+  const appController = React.useContext(AppControllerContext);
+
+  const [taskState, setTaskState] = React.useState<TaskState | null>(null);
+  const [taskLogs, setTaskLogs] = React.useState<TaskEventLog | null>(null);
+
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:35]:
```suggestion
  const [taskLogs, setTaskLogs] = React.useState<TaskEventLog | null>(null);
  const [isOpen, setIsOpen] = React.useState(true);

```

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+      appController.showAlert('error', `Failed to cancel task: ${(e as Error).message}`);
+    }
+  }, [appController, taskState, rmf]);
+
+  return (
+    <>
+      <Grid container wrap="nowrap" direction="column" height="100%">
+        <Dialog
+          open={open}
+          onClose={onClose}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:109]:
```suggestion
          open={isOpen}
          onClose={() => {
            setIsOpen(false);
            onClose();
          }}
```

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+          autoFocus
+        >
+          Inspect
+        </Button>
+      </DialogActions>
+      {openTaskDetailsLogs && (
+        <TaskInspector
+          task={taskState}
+          open={openTaskDetailsLogs}
+          onClose={() => setOpenTaskDetailsLogs(!openTaskDetailsLogs)}
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:204]:
```suggestion
          task={taskState}
          onClose={() => setOpenTaskDetailsLogs(false)}
```

**Review by aaronchongth — COMMENTED** [review]
As mentioned, this set of suggestions will remove the boolean prop and handle opening and closing.

File: packages/dashboard/src/components/tasks/task-summay.tsx [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+export const TaskSummary = React.memo((props: TaskSummaryProps) => {
+  const classes = useStyles();
+
+  const { stopShowing, show, task } = props;
+
+  const [openTaskDetailsLogs, setOpenTaskDetailsLogs] = React.useState(false);
+
+  const getTaskMessage = (task: TaskState | null) => {
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: True, I agree. Since we don't expect that much information from `Detail`. Let's pretty-print it with 2 spaces then

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+        ? Object.values(task.phases)[task.active - 1].category
+        : ''
+    }`;
+  };
+
+  const returnDialogContent = () => {
+    const contents = [
+      {
+        title: 'ID',
+        value: taskState ? taskState.booking.id : '',
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:138]:
```suggestion
        value: taskState ? taskState.booking.id : 'Invalid task state.',
```

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+      },
+      {
+        title: 'Current phase',
+        value: getTaskPhaseDetails(taskState),
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:142]:
```suggestion
        value: taskState ? getTaskPhaseDetails(taskState) : 'Invalid task state.',
```
We can probably follow your lead a few lines above and clean up the logic using a ternary

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+    if (!rmf || !task) {
+      return;
+    }
+    const sub = rmf
+      .getTaskStateObs(task.booking.id)
+      .subscribe((subscribedTask) => setTaskState(subscribedTask));
+    return () => sub.unsubscribe();
+  }, [rmf, task]);
+
+  const getTaskPhaseDetails = (task: TaskState | null) => {
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:112]:
```suggestion
  const getTaskPhaseDetails = (task: TaskState) => {
```
We can use non-null task states here now! :partying_face: Our error messages here can include the ID's safely now

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+    if (!task || !task.phases || !task.active) {
+      return 'Failed to retrieve current task phase';
+    }
+
+    if (!Object.values(task.phases)[task.active - 1]) {
+      return 'Failed to retrieve current task phase';
+    }
+
+    const message = Object.values(task.phases)[task.active - 1].detail;
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:121]:
```suggestion
    const message = Object.values(task.phases)[task.active - 1]?.detail;
```
We can further clean up the logic here

File: packages/dashboard/src/components/tasks/task-summary.tsx [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+
+    if (message) {
+      return message;
+    }
+
+    return `Failed to retrieve current task phase details of id ${task.booking.id} and category ${
+      Object.values(task.phases)[task.active - 1].category
+        ? Object.values(task.phases)[task.active - 1].category
+        : ''
+    }`;
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-summary.tsx:131]:
```suggestion
    const categoryString = Object.values(task.phases)[task.active - 1]?.category
      ? ` category ${Object.values(task.phases)[task.active - 1].category}`
      : '';
    return `Failed to retrieve current task phase details of id ${task.booking.id}${categoryString}`
```
We can really go crazy with ternaries here

File: packages/dashboard/src/components/tasks/task-summay.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: Just left some further comments! Resolving this one, thanks!

File: packages/dashboard/src/components/tasks/task-inspector.tsx [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+    }
+    const sub = rmf
+      .getTaskStateObs(task.booking.id)
+      .subscribe((subscribedTask) => setTaskState(subscribedTask));
+    return () => sub.unsubscribe();
+  }, [rmf, task]);
+
+  React.useEffect(() => {
+    AppEvents.taskSelect.next(task);
+  }, [task]);
```
Comment (aaronchongth) [inline packages/dashboard/src/components/tasks/task-inspector.tsx:74]: I don't think we need these lines anymore

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Just a few more nitpicks! But all is looking freaking amazing now, thanks for iterating on this!

File: packages/dashboard/src/components/tasks/task-inspector.tsx (hunk 3 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-inspector.tsx]: your changes were exactly what I was considering, thanks for implementing it so quickly! Resolved! :partying_face:

File: packages/dashboard/src/components/tasks/task-summay.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-summay.tsx]: I tried it doing this `message.toString().replace(/\s+/g, '   ');` but the result wasn't significant.  
Perhaps we can make the TextField higher with `maxRows=2`

File: packages/dashboard/src/components/tasks/task-inspector.tsx (hunk 27 above)
Comment (Angatupyry) [reply packages/dashboard/src/components/tasks/task-inspector.tsx]: You're right!

**Comment by Angatupyry** [comment]
> As mentioned, this set of suggestions will remove the boolean prop and handle opening and closing.

Oh, man! I misunderstood. I thought you meant removing even the onClose props. 
Thanks for taking the time to write this suggestion. This is definitely another alternative.

**Review by aaronchongth — APPROVED** [review]
Thanks for the hard work! This looks so good now! LGTM
